File: frontend/src/services/insights.ts
import { api } from './api'
import type { Insight } from '../types'

export const insightsService = {
  list: (params?: { limit?: number; resolved?: boolean; type?: string }) => {
    const search = new URLSearchParams()
    if (params?.limit != null) search.set('limit', String(params.limit))
    if (params?.resolved === true) search.set('resolved', 'true')
    if (params?.resolved === false) search.set('resolved', 'false')
    if (params?.type) search.set('type', params.type)
    const qs = search.toString()
    return api.get<Insight[]>(`/insights${qs ? `?${qs}` : ''}`)
  },
  run: () => api.post<Insight[]>('/insights/run', {}),
  resolve: (id: string) => api.patch<Insight>(`/insights/${id}/resolve`, {}),
}

File: backend/app/api/routes/insights.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.schemas.insight import InsightResponse
from app.services import insight_service

router = APIRouter()


@router.get("", response_model=list[InsightResponse])
def list_insights(
    limit: int = 50,
    resolved: bool | None = None,
    type: str | None = None,
    db: Session = Depends(get_db),
):
    """Get latest insights."""
    insights = insight_service.get_insights(db, limit=limit, resolved=resolved, type=type)
    return [InsightResponse.model_validate(i) for i in insights]


@router.post("/run", response_model=list[InsightResponse])
def run_insight_engine(db: Session = Depends(get_db)):
    """Trigger insight engine manually."""
    insights = insight_service.run_engine(db)
    return [InsightResponse.model_validate(i) for i in insights]


@router.patch("/{insight_id}/resolve", response_model=InsightResponse)
def resolve_insight(insight_id: str, db: Session = Depends(get_db)):
    """Mark insight as resolved."""
    insight = insight_service.resolve_insight(db, insight_id)
    if not insight:
        raise HTTPException(status_code=404, detail="Insight not found")
    return InsightResponse.model_validate(insight)
